File: 20_validParen.py
class Solution:
    def __init__(self) -> None:
        self.stack = []
    
    def match(self, a: str, b: str) -> bool:
        if a == '(' and b == ')':
            return True
        elif a == '[' and b == ']':
            return True
        elif a == '{' and b == '}':
            return True
        else:
            return False
    
    def isValid(self, s: str) -> bool:
        for i in s:
            if i in ['(', '[', '{']:
                self.stack.append(i)
            else:
                if self.stack != [] and self.match(self.stack[-1], i):
                    self.stack.pop()
                else:
                    return False
        return self.stack == []
    
S = Solution()
s = "([)]"
print(S.isValid(s))

File: 20_validParen.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<list>
#include<stack>
#include<stdlib.h>
using namespace std;

class Solution {
public:
    bool match(char a, char b){
        if(a == '(' && b == ')') return true;
        if(a == '[' && b == ']') return true;
        if(a == '{' && b == '}') return true;
        else return false;
    }

    bool isValid(string s) {
        if(s.length() == 1) return false;
        stack<char> st;
        for(int i = 0; i < s.size(); i++){
            if(s[i] == '(' || s[i] == '[' || s[i] == '{') st.push(s[i]);
            else {
                if(!st.empty() && match(st.top(), s[i])) st.pop();
                else return false;
            }
        }
        return st.empty();
    }
};

int main(){
    system("cls");
    // cout<<"It's SDA"<<endl;
    string s = "]";
    Solution S;
    cout << S.isValid(s) << endl;
    return 0;
}